Added in basic scripting features to Dolphin

Lobsterzelda opened the pull request:
This PR adds in basic scripting features to Dolphin.

It removes a lot of the features in my original scripting branch PR, in order to make the code easier to review (which will be added back in in future PRs).

Lobsterzelda commented:
@phire , @AdmiralCurtiss , @Pokechu22 , @JosJuice , @Hibyehello , @lioncash

This is a simplified version of my original large PR from here: https://github.com/dolphin-emu/dolphin/pull/11540

This branch is ready to be tested/merged right now. It designs+implements the APIs for the scripting DLLs to interact with Dolphin, creates the helper classes that are used by scripts, contains the code to handle running script callbacks (ex. OnInstructionHitCallbacks), and contains the code to initialize scripts from a DLL. Additionally, it also contains the Script Manager Window in order to start running scripts.

BhaaLseN commented on Source/Core/Core/HW/SI/SI_DeviceGCController.cpp:
```diff
@@ -153,6 +271,32 @@ GCPadStatus CSIDevice_GCController::GetPadStatus()
+  // Where GCControllerInputPolled callbacks are run when a movie is not being played back.
+  if (Scripting::ScriptUtilities::IsScriptingCoreInitialized() && !Movie::IsPlayingInput())
+  {
+    Scripting::OnGCControllerPolledCallbackAPI::current_game_cube_controller_number_polled =
```
That (and all the following members) are a mouthful to type...but I don't really have any better suggestion at the moment.

Feels like this might be better suited for methods (similar to `ScriptUtilities`) or maybe a class that encapsulates all of this (but I'm not really sure about performance and lifetime characteristics in relation to what _else_ you have planned here)

BhaaLseN commented on Source/Core/Core/PowerPC/MMU.cpp:
```diff
@@ -543,6 +546,30 @@ void MMU::Memcheck(u32 address, u64 var, bool write, size_t size)
+          value_written_to_memory_address_for_current_callback = var;
+      Scripting::OnMemoryAddressWrittenToCallbackAPI::write_size = static_cast<u32>(size);
+    }
+    if ((!write) && mc->is_break_on_read)
```
Do you have any particular reason why the `(!write)` condition is in parenthesis?

If not, the preferred style would be to omit them.
```suggestion
    if (!write && mc->is_break_on_read)
```

BhaaLseN commented on Source/Core/Core/Scripting/APIModules/EmuAPI.cpp:
```diff
@@ -0,0 +1,95 @@
+ArgHolder* EmuLoadState(ScriptContext* current_script, std::vector<ArgHolder*>* args_list)
+{
+  std::string load_state_name = (*args_list)[0]->string_val;
+  if (!CheckIfFileExists(load_state_name))
+    return CreateErrorStringArgHolder(
+        fmt::format("Could not find savestate with filename of {}", load_state_name).c_str());
```
```suggestion
  if (!CheckIfFileExists(load_state_name))
  {
    return CreateErrorStringArgHolder(
        fmt::format("Could not find savestate with filename of {}", load_state_name).c_str());
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressReadFromCallbackAPI.cpp:
```diff
@@ -0,0 +1,153 @@
+  u32 memory_breakpoint_end_address = (*args_list)[1]->u32_val;
+  void* callback = (*args_list)[2]->void_pointer_val;
+
+  if (memory_breakpoint_start_address == 0)
+    return CreateErrorStringArgHolder("Error: Memory address read breakpoint cannot be 0!");
+  else if (memory_breakpoint_start_address > memory_breakpoint_end_address)
+    return CreateErrorStringArgHolder("Error: Memory address read breakpoint has ending address "
+                                      "less than its starting address!");
```
```suggestion
  if (memory_breakpoint_start_address == 0)
  {
    return CreateErrorStringArgHolder("Error: Memory address read breakpoint cannot be 0!");
  }
  else if (memory_breakpoint_start_address > memory_breakpoint_end_address)
  {
    return CreateErrorStringArgHolder("Error: Memory address read breakpoint has ending address "
                                      "less than its starting address!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressReadFromCallbackAPI.cpp:
```diff
@@ -0,0 +1,153 @@
+  u32 memory_breakpoint_start_address = (*args_list)[0]->u32_val;
+  void* callback = (*args_list)[1]->void_pointer_val;
+
+  if (!current_script->memory_address_breakpoints_holder.ContainsReadBreakpoint(
+          memory_breakpoint_start_address))
+    return CreateErrorStringArgHolder(
+        "Error: Address passed into OnMemoryAddressReadFrom.Unregister() did not represent the "
+        "start address of a read "
+        "breakpoint that was currently enabled!");
```
```suggestion
  if (!current_script->memory_address_breakpoints_holder.ContainsReadBreakpoint(
          memory_breakpoint_start_address))
  {
    return CreateErrorStringArgHolder(
        "Error: Address passed into OnMemoryAddressReadFrom.Unregister() did not represent the "
        "start address of a read "
        "breakpoint that was currently enabled!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressReadFromCallbackAPI.cpp:
```diff
@@ -0,0 +1,153 @@
+      current_script->dll_specific_api_definitions.UnregisterOnMemoryAddressReadFromCallback(
+          current_script, callback);
+
+  if (!return_value)
+    return CreateErrorStringArgHolder(
+        "2nd Argument passed into OnMemoryAddressReadFrom.unregister() was not a reference to a "
+        "function currently registered as an OnMemoryAddressReadFrom callback!");
+  else
+    return CreateVoidTypeArgHolder();
```
```suggestion
  if (!return_value)
  {
    return CreateErrorStringArgHolder(
        "2nd Argument passed into OnMemoryAddressReadFrom.unregister() was not a reference to a "
        "function currently registered as an OnMemoryAddressReadFrom callback!");
  }
  else
  {
    return CreateVoidTypeArgHolder();
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressReadFromCallbackAPI.cpp:
```diff
@@ -0,0 +1,153 @@
+ArgHolder* GetMemoryAddressReadFromForCurrentCallback(ScriptContext* current_script,
+                                                      std::vector<ArgHolder*>* args_list)
+{
+  if (current_script->current_script_call_location !=
+      ScriptCallLocationsEnum::FromMemoryAddressReadFromCallback)
+    return CreateErrorStringArgHolder(
+        "User attempted to call "
+        "OnMemoryAddressReadFrom.getMemoryAddressReadFromForCurrentCallback() outside of an "
+        "OnMemoryAddressReadFrom callback function!");
```
```suggestion
  if (current_script->current_script_call_location !=
      ScriptCallLocationsEnum::FromMemoryAddressReadFromCallback)
  {
    return CreateErrorStringArgHolder(
        "User attempted to call "
        "OnMemoryAddressReadFrom.getMemoryAddressReadFromForCurrentCallback() outside of an "
        "OnMemoryAddressReadFrom callback function!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressReadFromCallbackAPI.cpp:
```diff
@@ -0,0 +1,153 @@
+
+ArgHolder* GetReadSize(ScriptContext* current_script, std::vector<ArgHolder*>* args_list)
+{
+  if (current_script->current_script_call_location !=
+      ScriptCallLocationsEnum::FromMemoryAddressReadFromCallback)
+    return CreateErrorStringArgHolder("User attempted to call "
+                                      "OnMemoryAddressReadFrom.getReadSize() outside of an "
+                                      "OnMemoryAddressReadFrom callback function!");
```
```suggestion
  if (current_script->current_script_call_location !=
      ScriptCallLocationsEnum::FromMemoryAddressReadFromCallback)
  {
    return CreateErrorStringArgHolder("User attempted to call "
                                      "OnMemoryAddressReadFrom.getReadSize() outside of an "
                                      "OnMemoryAddressReadFrom callback function!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressWrittenToCallbackAPI.cpp:
```diff
@@ -0,0 +1,172 @@
+  u32 memory_breakpoint_end_address = (*args_list)[1]->u32_val;
+  void* callback = (*args_list)[2]->void_pointer_val;
+
+  if (memory_breakpoint_start_address == 0)
+    return CreateErrorStringArgHolder("Error: Memory address write breakpoint cannot be 0!");
+  else if (memory_breakpoint_start_address > memory_breakpoint_end_address)
+    return CreateErrorStringArgHolder(
+        "Error: Memory address write breakpoint has an ending address less "
+        "than its starting address!");
```
```suggestion
  if (memory_breakpoint_start_address == 0)
  {
    return CreateErrorStringArgHolder("Error: Memory address write breakpoint cannot be 0!");
  }
  else if (memory_breakpoint_start_address > memory_breakpoint_end_address)
  {
    return CreateErrorStringArgHolder(
        "Error: Memory address write breakpoint has an ending address less "
        "than its starting address!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressWrittenToCallbackAPI.cpp:
```diff
@@ -0,0 +1,172 @@
+  u32 memory_breakpoint_end_address = (*args_list)[1]->u32_val;
+  void* callback = (*args_list)[2]->void_pointer_val;
+
+  if (memory_breakpoint_start_address == 0)
+    return CreateErrorStringArgHolder("Error: Memory address write breakpoint cannot be 0!");
+  else if (memory_breakpoint_start_address > memory_breakpoint_end_address)
+    return CreateErrorStringArgHolder("Error: Memory address write breakpoint has an ending "
+                                      "address less than its starting address!");
```
```suggestion
  if (memory_breakpoint_start_address == 0)
  {
    return CreateErrorStringArgHolder("Error: Memory address write breakpoint cannot be 0!");
  }
  else if (memory_breakpoint_start_address > memory_breakpoint_end_address)
  {
    return CreateErrorStringArgHolder("Error: Memory address write breakpoint has an ending "
                                      "address less than its starting address!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressWrittenToCallbackAPI.cpp:
```diff
@@ -0,0 +1,172 @@
+  bool return_value =
+      current_script->dll_specific_api_definitions.UnregisterOnMemoryAddressWrittenToCallback(
+          current_script, callback);
+  if (!return_value)
+    return CreateErrorStringArgHolder(
+        "2nd argument passed into OnMemoryAddressWrittenTo.unregister() was not a reference to a "
+        "function currently registered as an OnMemoryAddressWrittenTo callback!");
+  else
+    return CreateVoidTypeArgHolder();
```
```suggestion
  if (!return_value)
  {
    return CreateErrorStringArgHolder(
        "2nd argument passed into OnMemoryAddressWrittenTo.unregister() was not a reference to a "
        "function currently registered as an OnMemoryAddressWrittenTo callback!");
  }
  else
  {
    return CreateVoidTypeArgHolder();
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressWrittenToCallbackAPI.cpp:
```diff
@@ -0,0 +1,172 @@
+ArgHolder* GetMemoryAddressWrittenToForCurrentCallback(ScriptContext* current_script,
+                                                       std::vector<ArgHolder*>* args_list)
+{
+  if (current_script->current_script_call_location !=
+      ScriptCallLocationsEnum::FromMemoryAddressWrittenToCallback)
+    return CreateErrorStringArgHolder(
+        "User attempted to call "
+        "OnMemoryAddressWrittenTo.getMemoryAddressWrittenToForCurrentCallback() outside of an "
+        "OnMemoryAddressWrittenTo callback function!");
```
```suggestion
  if (current_script->current_script_call_location !=
      ScriptCallLocationsEnum::FromMemoryAddressWrittenToCallback)
  {
    return CreateErrorStringArgHolder(
        "User attempted to call "
        "OnMemoryAddressWrittenTo.getMemoryAddressWrittenToForCurrentCallback() outside of an "
        "OnMemoryAddressWrittenTo callback function!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressWrittenToCallbackAPI.cpp:
```diff
@@ -0,0 +1,172 @@
+ArgHolder* GetValueWrittenToMemoryAddressForCurrentCallback(ScriptContext* current_script,
+                                                            std::vector<ArgHolder*>* args_list)
+{
+  if (current_script->current_script_call_location !=
+      ScriptCallLocationsEnum::FromMemoryAddressWrittenToCallback)
+    return CreateErrorStringArgHolder(
+        "User attempted to call "
+        "OnMemoryAddressWrittenTo.getValueWrittenToMemoryAddressForCurrentCallback() outside of an "
+        "OnMemoryAddressWrittenTo callback function!");
```
```suggestion
  if (current_script->current_script_call_location !=
      ScriptCallLocationsEnum::FromMemoryAddressWrittenToCallback)
  {
    return CreateErrorStringArgHolder(
        "User attempted to call "
        "OnMemoryAddressWrittenTo.getValueWrittenToMemoryAddressForCurrentCallback() outside of an "
        "OnMemoryAddressWrittenTo callback function!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnMemoryAddressWrittenToCallbackAPI.cpp:
```diff
@@ -0,0 +1,172 @@
+
+ArgHolder* GetWriteSize(ScriptContext* current_script, std::vector<ArgHolder*>* args_list)
+{
+  if (current_script->current_script_call_location !=
+      ScriptCallLocationsEnum::FromMemoryAddressWrittenToCallback)
+    return CreateErrorStringArgHolder("User attempted to call "
+                                      "OnMemoryAddressWrittenTo.getWriteSize() outside of an "
+                                      "OnMemoryAddressWrittenTo callback function!");
```
```suggestion
  if (current_script->current_script_call_location !=
      ScriptCallLocationsEnum::FromMemoryAddressWrittenToCallback)
  {
    return CreateErrorStringArgHolder("User attempted to call "
                                      "OnMemoryAddressWrittenTo.getWriteSize() outside of an "
                                      "OnMemoryAddressWrittenTo callback function!");
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/EventCallbackRegistrationAPIs/OnWiiInputPolledCallbackAPI.cpp:
```diff
@@ -0,0 +1,78 @@
+  bool return_value =
+      current_script->dll_specific_api_definitions.UnregisterOnWiiInputPolledCallback(
+          current_script, (*args_list)[0]->void_pointer_val);
+  if (!return_value)
+    return CreateErrorStringArgHolder(
+        "Argument passed into OnWiiInputPolled.unregister() was not a reference to a function "
+        "currently registered as an OnWiiInputPolled callback!");
+  else
+    return CreateVoidTypeArgHolder();
```
```suggestion
  if (!return_value)
  {
    return CreateErrorStringArgHolder(
        "Argument passed into OnWiiInputPolled.unregister() was not a reference to a function "
        "currently registered as an OnWiiInputPolled callback!");
  }
  else
  {
    return CreateVoidTypeArgHolder();
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/ArgHolder.cpp:
```diff
@@ -0,0 +1,237 @@
+ArgHolder* CreateEmptyOptionalArgument()
+{
+  ArgHolder* return_val = new ArgHolder();
+  return_val->contains_value =
+      false;  // We always check if the ArgHolder contains a value before looking at anything else.
```
This probably reads nicer if you put the comment on the line before.
```suggestion
  // We always check if the ArgHolder contains a value before looking at anything else.
  return_val->contains_value = false;
```

Lobsterzelda replied:
Fixed.

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/ArgHolder.h:
```diff
@@ -0,0 +1,87 @@
+struct ArgHolder
+{
+  ArgTypeEnum argument_type;
+  bool contains_value;  // True if the ArgHolder contains a value, and false if it's empty.
```
If you put this before the member (rather than at the end of the line), it also shows up in various IDEs as documentation.
```suggestion
  // True if the ArgHolder contains a value, and false if it's empty.
  bool contains_value;
```

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/ClassFunctionsResolver_API_Implementations.cpp:
```diff
@@ -0,0 +1,128 @@
+      GetClassMetadataForModule(std::string(module_name), std::string(version_number));
+  reinterpret_cast<ScriptContext*>(script_context)
+      ->dll_specific_api_definitions.DLLClassMetadataCopyHook(script_context, &classMetadata);
+  classMetadata.class_name =
+      classMetadata
+          .class_name;  // ... just putting this here to prevent the compiler from trying to delete
+                        // classMetadata early (although I'm 99% sure that's impossible).
```
This (and the one in the next method) seems really janky. I'd be surprised if this didn't trigger at least one compiler warning _somewhere_ for a pointless or mistaken assignment (or be optimized out anyways)

Feels like it shouldn't be necessary (unless `DLLClassMetadataCopyHook` returns early and continues running on a different thread)?

Lobsterzelda replied:
Removed this.

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/FunctionMetadata.h:
```diff
@@ -0,0 +1,63 @@
+public:
+  FunctionMetadata()
+  {
+    this->function_name = "";
+    this->function_version = "";
+    this->example_function_call = "";
+    function_pointer = nullptr;
+    return_type = (ArgTypeEnum)0;
+    arguments_list = std::vector<ArgTypeEnum>();
```
This (and the following ctor) seem really inconsistent in style. I don't think we made any decision either way in our code style; but the prevailing style elsewhere is _without_ the `this->` qualification (unless it is necessary to make the unabmiguous; which shouldn't ever be necessary since we prefer classes to use the `m_` prefix for members)

Lobsterzelda replied:
Fixed.

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/GCButton_API_Implementations.cpp:
```diff
@@ -0,0 +1,292 @@
+namespace Scripting
+{
+
+static bool IsEqualIgnoreCase(const char* string_1, const char* string_2)
+{
+  int index = 0;
+  while (string_1[index] != '\0')
+  {
+    if (string_1[index] != string_2[index])
+    {
+      char first_char = string_1[index];
+      if (first_char >= 65 && first_char <= 90)
+        first_char += 32;
+      char second_char = string_2[index];
+      if (second_char >= 65 && second_char <= 90)
+        second_char += 32;
+      if (first_char != second_char)
+        return false;
+    }
+    ++index;
+  }
+
+  return string_2[index] == '\0';
+}
```
Feels odd that you _have to_ implement this by hand. What about using `strcasecmp` and its friends?

Depending on _where else_ you need this, we got `StringUtil::CaseInsensitiveEquals`, but it takes a `string_view` as parameter (and not a `char*`)

Lobsterzelda replied:
@BhaaLseN This is the one change you suggested that I couldn't resolve, since strcasecmp isn't a standard C++ function (and isn't available from Dolphin by default without adding in other imports).

noahpistilli replied:
`const char*` should be able to be passed as an `std::string_view`.

Lobsterzelda replied:
Added in those change, so this issue is now resolved.

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/GCButton_API_Implementations.cpp:
```diff
@@ -0,0 +1,292 @@
+    }
+  }
+
+  switch (button_name[0])
```
This doesn't verify that `button_name` is actually set. You check for length 1 earlier, but never for 0.

And since this is user-controlled, it should be a bit more sturdy than that (unless I missed an earlier place that doesn't let you call this with an empty string)

Lobsterzelda replied:
Fixed.

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/InstructionBreakpointsHolder.cpp:
```diff
@@ -0,0 +1,53 @@
+
+void InstructionBreakpointsHolder::AddBreakpoint(unsigned int addr)
+{
+  if (!this->ContainsBreakpoint(addr))
+    Core::System::GetInstance().GetPowerPC().GetBreakPoints().Add(addr, false, true, false,
+                                                                  std::nullopt);
```
```suggestion
  if (!this->ContainsBreakpoint(addr))
  {
    Core::System::GetInstance().GetPowerPC().GetBreakPoints().Add(addr, false, true, false,
                                                                  std::nullopt);
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/MemoryAddressBreakpointsHolder.cpp:
```diff
@@ -0,0 +1,161 @@
+void MemoryAddressBreakpointsHolder::RemoveWriteBreakpoint(unsigned int start_address)
+{
+  int index_of_last_write_breakpoint = GetIndexOfStartAddress(start_address);
+  if (index_of_last_write_breakpoint == -1 ||
+      memory_breakpoints[index_of_last_write_breakpoint].write_breakpoint_count == 0)
+    return;
```
```suggestion
  if (index_of_last_write_breakpoint == -1 ||
      memory_breakpoints[index_of_last_write_breakpoint].write_breakpoint_count == 0)
  {
    return;
  }
```

BhaaLseN commented on Source/Core/Core/Scripting/HelperClasses/VersionResolver.h:
```diff
@@ -0,0 +1,94 @@
+    std::string current_function_name = all_functions[i].function_name;
+    std::string function_version_number = all_functions[i].function_version;
+
+    if (function_to_latest_version_found_map.count(current_function_name) == 0 &&
+        !IsFirstVersionGreaterThanSecondVersion(function_version_number, api_version))
+      function_to_latest_version_found_map[current_function_name] = all_functions[i];
+
+    else if (function_to_latest_version_found_map.count(current_function_name) > 0 &&
+             IsFirstVersionGreaterThanSecondVersion(
+                 function_version_number,
+                 function_to_latest_version_found_map[current_function_name].function_version) &&
+             !IsFirstVersionGreaterThanSecondVersion(function_version_number, api_version))
+      function_to_latest_version_found_map[current_function_name] = all_functions[i];
```
```suggestion
    if (function_to_latest_version_found_map.count(current_function_name) == 0 &&
        !IsFirstVersionGreaterThanSecondVersion(function_version_number, api_version))
    {
      function_to_latest_version_found_map[current_function_name] = all_functions[i];
    }
    else if (function_to_latest_version_found_map.count(current_function_name) > 0 &&
             IsFirstVersionGreaterThanSecondVersion(
                 function_version_number,
                 function_to_latest_version_found_map[current_function_name].function_version) &&
             !IsFirstVersionGreaterThanSecondVersion(function_version_number, api_version))
    {
      function_to_latest_version_found_map[current_function_name] = all_functions[i];
    }
```

Lobsterzelda commented:
@BhaaLseN I cut out the EventCallbackRegistration directory entirely, removed EmuAPI, removed most references to Scripts from outside of the Script directory, and replaced some of the functions in ScriptUtilities.cpp with TODOs in order to cut down on the amount of code in this PR.

This has resulted in the PR going from about 7,800 changed lines to about 4,800 changed lines.

I'm hopeful that this will make reviewing the PR a bit more manageable.

BhaaLseN commented:
It certainly helps, but wasn't my intention to have you cut things out; i mainly meant that _I_ need another pass over it once I cleared my head. Which is hopefully this weekend, but no promises.
But if it makes sense as a smaller PR, by all means go for it.

Lobsterzelda commented:
> It certainly helps, but wasn't my intention to have you cut things out; i mainly meant that _I_ need another pass over it once I cleared my head. Which is hopefully this weekend, but no promises. But if it makes sense as a smaller PR, by all means go for it.

It's all good. I was planning to make the PR smaller anyways.

@BhaaLseN Btw, I finished making all of the changes you suggested.

lioncash commented on Source/Core/Core/Scripting/ScriptUtilities.cpp:
```diff
@@ -0,0 +1,764 @@
+
+bool StartScripts()
+{
+  std::lock_guard<std::mutex> lock(initialization_and_destruction_lock);
```
If you're acquiring two mutexes in a row, you may want to do

```cpp
std::scoped_lock lock(initialization_and_destruction_lock, global_code_and_frame_callback_running_lock);
```

which functions like an in-place `std::lock`, preventing deadlocks

lioncash commented on Source/Core/Core/Scripting/ScriptUtilities.cpp:
```diff
@@ -0,0 +1,764 @@
+
+bool RunGlobalCode()
+{
+  std::lock_guard<std::mutex> lock(global_code_and_frame_callback_running_lock);
```
```suggestion
  std::lock_guard lock(global_code_and_frame_callback_running_lock);
```

lioncash commented on Source/Core/Core/Scripting/ScriptUtilities.cpp:
```diff
@@ -0,0 +1,764 @@
+
+bool RunOnFrameStartCallbacks()
+{
+  std::lock_guard<std::mutex> lock(global_code_and_frame_callback_running_lock);
```
```suggestion
  std::lock_guard lock(global_code_and_frame_callback_running_lock);
```

ditto for other locks in the file

lioncash commented on Source/Core/DolphinQt/ScriptManager.cpp:
```diff
@@ -0,0 +1,228 @@
+  m_load_script_button = new NonDefaultQPushButton(tr("Load Script"));
+
+  m_play_or_stop_script_button = new NonDefaultQPushButton(tr("N/A"));
+  m_play_or_stop_script_button->setStyleSheet(tr("background-color:grey;"));
```
```suggestion
  m_play_or_stop_script_button->setStyleSheet(QStringLiteral("background-color:grey;"));
```

This doesn't need to be translated

lioncash commented on Source/Core/DolphinQt/ScriptManager.cpp:
```diff
@@ -0,0 +1,228 @@
+  output_box->addWidget(script_output_list_widget_ptr, 1, 0, Qt::AlignTop);
+
+  script_output_list_widget_ptr->setSpacing(1);
+  script_output_list_widget_ptr->setStyleSheet(tr("background-color:white;"));
```
```suggestion
  script_output_list_widget_ptr->setStyleSheet(QStringLiteral("background-color:white;"));
```

lioncash commented on Source/Core/DolphinQt/ScriptManager.cpp:
```diff
@@ -0,0 +1,228 @@
+
+void ScriptManager::PlayScriptFunction()
+{
+  print_lock.lock();
```
It's generally more idiomatic to use a block scope for delineating locking boundaries with a RAII guard:

```cpp
{
  std::lock_guard lock(print_lock);
  // contents here
}
```

lioncash commented on Source/Core/DolphinQt/ScriptManager.cpp:
```diff
@@ -0,0 +1,228 @@
+  output_lines.clear();
+  print_lock.unlock();
+
+  script_start_or_stop_lock.lock();
```
Ditto

lioncash commented on Source/Core/DolphinQt/ScriptManager.cpp:
```diff
@@ -0,0 +1,228 @@
+    row_num_to_is_running[next_unique_identifier] = false;
+    QStringList list;
+    list << path;
+    script_name_list_widget_ptr->addItem(path);
```
```suggestion
    script_name_list_widget_ptr->addItem(QStringList(path));
```

is more compact compared to constructing the list and manually appending to it.

lioncash commented on Source/Core/DolphinQt/ScriptManager.cpp:
```diff
@@ -0,0 +1,228 @@
+    UpdateButtonText();
+  }
+
+  return;
```
This can be omitted

lioncash commented on Source/Core/Core/Scripting/CoreScriptInterface/InternalScriptAPIs/ArgHolder_APIs.h:
```diff
@@ -0,0 +1,188 @@
+  void* (*CreateBoolArgHolder)(int);
+
+  // Creates + returns an unsigned_8 ArgHolder* with the specified value.
+  void* (*CreateU8ArgHolder)(unsigned long long);
```
maybe just use uint64_t, etc for the relevant functions?

Lobsterzelda replied:
For all of the function pointers which are going to be passed across the DLL boundary (such as this one), I tried to make all integer types in the signature/parameters either unsigned char, signed char, unsigned long long, and signed long long. This way, I could be sure that the data I'm passing would be either 1 byte or the maximum bytesize for the system (realistically, 8 bytes for any computer running Dolphin).

Also, I want to make sure that the type exists in plain C, which is why I tried to avoid including other types (ex. I return ints from functions instead of bool, since C doesn't have a bool type).

ElectrifiedStrawberry replied:
Sorry to necro but one thing is that using variable length types I think causes annoying ABI concerns (especially across languages that may not have ULL as an inbuilt type), While ULL (`unsigned long long`) is guaranteed to be 64 bits or larger, this makes things incompatible if Dolphin is rebuilt and ULL changes.

And realistically if something doesn't have `uint64_t`, it's already a super esoteric implementation of C in the first place.

lioncash commented on Source/Core/Core/Scripting/CoreScriptInterface/InternalScriptAPIs/ArgHolder_APIs.h:
```diff
@@ -0,0 +1,188 @@
+  void* (*CreateBytesArgHolder)();
+
+  // Takes a std::vector<s16> ArgHolder* as input, and returns the number of bytes in the list.
+  unsigned long long (*GetSizeOfBytesArgHolder)(void*);
```
```suggestion
  size_t (*GetSizeOfBytesArgHolder)(void*);
```

lioncash commented on Source/Core/Core/Scripting/HelperClasses/FunctionMetadata.cpp:
```diff
@@ -0,0 +1,67 @@
+namespace Scripting
+{
+
+FunctionMetadata* castToFunctionMetadataPtr(void* input)
```
```suggestion
static FunctionMetadata* CastToFunctionMetadataPtr(void* input)
```

Lobsterzelda commented on Source/Core/Core/Scripting/HelperClasses/FileUtility_API_Implementations.cpp:
```diff
@@ -0,0 +1,28 @@
+std::string user_path = "";
+std::string sys_path = "";
+
+void SetUserPath(const std::string new_user_path)
```
Wow! This is a very thorough list of suggestions/fixes. Thank you for taking the time to go through all of my code! Btw, I was wondering, what's the reason behind getting rid of the const qualifier for the parameter here?

JosJuice replied:
So that you can std::move it.

Lobsterzelda commented on Source/Core/Core/Scripting/HelperClasses/ScriptContext.h:
```diff
@@ -0,0 +1,92 @@
+namespace Scripting
+{
+
+typedef struct ScriptContext
```
What is the benefit of removing the typedef here?

JosJuice commented on Source/Core/Core/Scripting/HelperClasses/GCButton_API_Implementations.cpp:
```diff
@@ -0,0 +1,279 @@
+int ParseGCButton_impl(const char* button_name)
+{
+  if (button_name == nullptr)
+    return (int)GCButtonNameEnum::UnknownButton;
```
Either that or `static_cast`.

Lobsterzelda replied:
I refactored this file to use static_cast for everything.

Lobsterzelda commented:
@lioncash I added in all of the fixes you suggested, and marked them as resolved. The only 2 changes I didn't do were the ones in ArgHolder_APIs.h, since I want to keep those functions C-compatible.

Lobsterzelda commented:
I updated the PR to make it even shorter, so that it would be easier to read. I also added 2 new README type files inside of the Source/Core/Core/Scripting folder:

https://github.com/Lobsterzelda/lobster-zelda-dolphin/blob/bare-bones-scripting-branch/Source/Core/Core/Scripting/ScriptingPlugins_Documentation.md

ScriptingPlugins_Documentation.md describes how to create a new ScriptingPlugin (to add in support for a new Scripting language), and describes what functions a DLL need to export/implement. It also describes the general process for adding in a new scripting language.

https://github.com/Lobsterzelda/lobster-zelda-dolphin/blob/bare-bones-scripting-branch/Source/Core/Core/Scripting/Common_Scripting_Functions_Documentation.md

Common_Scripting_Functions_Documentation.md describes what the purpose is behind each file in this PR.

Both of these files are about 200 lines, so I'm hoping that this short documentation will make it easier to read+understand the components of this PR (which will hopefully lead to this PR being merged).

@phire , @Pokechu22 , @AdmiralCurtiss , @Hibyehello , @lioncash , @BhaaLseN , @JosJuice , @CasualPokePlayer

BhaaLseN commented on Source/Core/Core/Scripting/ScriptingPlugins_Documentation.md:
```diff
@@ -0,0 +1,226 @@
+
+As an example, suppose we want to implement the Lua language. Then, our file structure will look like this, with the root directory representing the root directory for Dolphin (D represents directories, F represents files):
+
+/ (D)
+
+----PluginSources (D)
+
+--------Lua (D)
+
+------------extensions.txt (F)
+
+------------Source (D)
+
+----------------CoreScriptInterface (D)
+
+----------------<The source code for the DLL> (Multiple files and or directories).
```
This seems difficult to read, how about making this a code block instead?
```suggestion
[/]
+- [PluginSources]
   +- [Lua]
      +- extensions.txt
      +- [Source]
         +- [CoreScriptInterface]
         +- <The source code for the DLL> (Multiple files and or directories).
```

Not really feeling that one either, but maybe we can come up with something more readable together.

BhaaLseN reviewed (commented):
I just skimmed over this; and I generally prefer `identifiers` to be marked up (or rather, down, using single backticks `); but I also see that this might be quite the effort, so I won't insist on getting that done to the two markdown files.

JosJuice commented on Source/Core/Core/Scripting/HelperClasses/InstructionBreakpointsHolder.cpp:
```diff
@@ -0,0 +1,52 @@
+
+bool InstructionBreakpointsHolder::ContainsBreakpoint(unsigned int addr) const
+{
+  return (std::count(breakpoint_addresses.begin(), breakpoint_addresses.end(), addr) > 0);
```
C++20 should be fine nowadays.

twitnic commented:
Any Update for this PR? Lua Support / a art of scripting would be really nice ...

Lobsterzelda commented:
> Any Update for this PR? Lua Support / a art of scripting would be really nice ...

Basically, the only person with both the time and scripting knowledge/experience to review this MR would be @phire 

However, Phire hasn't been online for the last 6 or so months, so it's unclear when this MR will be reviewed+approved.

TryTwo commented:
Is there a full build with python that I can mess around with somewhere?

Lobsterzelda commented:
> Is there a full build with python that I can mess around with somewhere?

Yes. You can use this project code: https://github.com/Lobsterzelda/lobster-zelda-dolphin/tree/lua-support

InputEvelution commented:
> Basically, the only person with both the time and scripting knowledge/experience to review this MR would be @phire
> However, Phire hasn't been online for the last 6 or so months, so it's unclear when this MR will be reviewed+approved.

Any update on this? It would be a shame if this never got merged because a single Dolphin developer isn't active enough to look at it...

Lobsterzelda commented:
> > Basically, the only person with both the time and scripting knowledge/experience to review this MR would be @phire
> > However, Phire hasn't been online for the last 6 or so months, so it's unclear when this MR will be reviewed+approved.
> 
> Any update on this? It would be a shame if this never got merged because a single Dolphin developer isn't active enough to look at it...

Sorry for the late reply - I just saw this post.

Basically, this is probably never going to be merged. It isn't just that Phire isn't around to merge this. In order for this to be merged, it would probably require ~5 (or maybe more) extremely highly-ranked developers for Dolphin who have the ability to understand the scripting MR and who also have Dolphin merge permission to decide to commit more than a month to reviewing/testing this. As it stands, there are currently 0 dolphin developers who meet both criteria (i.e. those who are ranked highly enough to be able to merge this either don't understand scripting well enough to be able to review this, or they don't want to invest months into adding in this feature, and those who do have the knowledge to review/test this MR don't rank highly enough to be able to merge anything).

I've given up hope on this ever being merged in any capacity. However, I am creating a new scripting branch which will be compatible with modern builds of Dolphin, since I want to make a TAS of a game on a modern version of Dolphin, and having access to my scripting branch will be helpful for that purpose. Other TASers can also use my branch when it's done, but this is probably going to end up suffering the same fate as every other Dolphin scripting project that's ever been attempted - relegated to a permanent Dolphin fork, which will never be merged into the actual Dolphin repository...
